Source/SwiftLintFramework/Rules/Lint/DuplicatedKeyInDictionaryLiteralRule.swift

```diff
@@ -79,74 +79,59 @@ public struct DuplicatedKeyInDictionaryLiteralRule: ASTRule, ConfigurationProvid
     )
 
-    public func validate(file: SwiftLintFile, kind: SwiftExpressionKind,
-                         dictionary: SourceKittenDictionary) -> [StyleViolation] {
-        guard kind == .dictionary else {
-            return []
-        }
-
-        let keys = nonGeneratedDictionaryKeys(with: file, dictionary: dictionary)
-        guard keys.count >= 2 else {
-            return []
-        }
+    public func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
 
-        var existingKeys: [String: DictionaryKey] = [:]
-        return keys
-            .filter { key in
-                guard let existingKey = existingKeys[key.content] else {
-                    existingKeys[key.content] = key
-                    return false
+private extension DuplicatedKeyInDictionaryLiteralRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ list: DictionaryElementListSyntax) {
+            let keys = list.map(\.keyExpression).compactMap { expr -> DictionaryKey? in
+                expr.stringContent.map {
+                    DictionaryKey(position: expr.positionAfterSkippingLeadingTrivia, content: $0)
                 }
-
-                let existingKeyKinds = file.syntaxMap.kinds(inByteRange: existingKey.byteRange)
-                let keyKinds = file.syntaxMap.kinds(inByteRange: key.byteRange)
-                return keyKinds == existingKeyKinds
-            }.map { key in
-                StyleViolation(ruleDescription: Self.description,
-                               severity: configuration.severity,
-                               location: Location(file: file, byteOffset: key.byteRange.location))
             }
-    }
 
-    private func nonGeneratedDictionaryKeys(with file: SwiftLintFile,
-                                            dictionary: SourceKittenDictionary) -> [DictionaryKey] {
-        let keys = dictionary.elements.enumerated().compactMap { index, element -> SourceKittenDictionary? in
-            // in a dictionary, the even elements are keys, and the odd elements are values
-            if index.isMultiple(of: 2) {
-                return element
+            guard keys.count >= 2 else {
+                return
             }
-            return nil
-        }.filter {
-            guard let key = $0.content(in: file) else { return true }
-            return !isCodeGeneratedKey(keyExpression: key)
-        }
 
-        let contents = file.stringView
-        return keys.compactMap { key -> DictionaryKey? in
-            guard let range = key.byteRange,
-                  let substring = contents.substringWithByteRange(range) else {
-                return nil
-            }
+            var existingKeys: [String: DictionaryKey] = [:]
+            let newViolations = keys
+                .filter { key in
+                    guard existingKeys[key.content] != nil else {
+                        existingKeys[key.content] = key
```

Comment:
Nit, personal opinion: this mutation side-effect doesn't belong in a `filter`

Reply:
updated!